lntest/itest/lnd_psbt_test.go

```diff
@@ -727,6 +734,182 @@ func runSignPsbtSegWitV0P2WKH(t *harnessTest, net *lntest.NetworkHarness,
+
+	// Everything we do here should be done within a second or two, so we
+	// can just keep a single timeout context around for all calls.
+	ctxb := context.Background()
+	ctxt, cancel := context.WithTimeout(ctxb, defaultTimeout)
+	defer cancel()
+
+	// Derive a key we can use for signing.
+	keyDesc, internalKey, fullDerivationPath := deriveInternalKey(
+		ctxt, t, alice,
+	)
+
+	// Let's create a taproot script output now. This is a hash lock with a
+	// simple preimage of "foobar".
+	leaf1 := testScriptHashLock(t.t, []byte("foobar"))
+
+	rootHash := leaf1.TapHash()
+	taprootKey := txscript.ComputeTaprootOutputKey(internalKey, rootHash[:])
+	tapScriptAddr, err := btcutil.NewAddressTaproot(
+		schnorr.SerializePubKey(taprootKey), harnessNetParams,
+	)
+	require.NoError(t.t, err)
+	p2trPkScript, err := txscript.PayToAddrScript(tapScriptAddr)
+	require.NoError(t.t, err)
+
+	// Send some funds to the output and then try to get a signature through
+	// the SignPsbt RPC to spend that output again.
+	assertPsbtSpend(
+		ctxt, t, net, alice, p2trPkScript,
+		func(packet *psbt.Packet) {
+			in := &packet.Inputs[0]
+			in.Bip32Derivation = []*psbt.Bip32Derivation{{
+				PubKey:    keyDesc.RawKeyBytes,
+				Bip32Path: fullDerivationPath,
+			}}
+			in.TaprootBip32Derivation = []*psbt.TaprootBip32Derivation{{
+				XOnlyPubKey: keyDesc.RawKeyBytes[1:],
+				Bip32Path:   fullDerivationPath,
+			}}
+			in.TaprootMerkleRoot = rootHash[:]
+			in.SighashType = txscript.SigHashDefault
+		},
+		func(packet *psbt.Packet) {
+			require.Len(t.t, packet.Inputs, 1)
+			require.Len(
+				t.t, packet.Inputs[0].TaprootKeySpendSig, 64,
+			)
+		},
+	)
+}
+
+// runSignPsbtSegWitV1ScriptSpend tests that the SignPsbt RPC works correctly
+// for a SegWit v1 p2tr script spend.
+func runSignPsbtSegWitV1ScriptSpend(t *harnessTest,
+	net *lntest.NetworkHarness, alice *lntest.HarnessNode) {
+
+	// Everything we do here should be done within a second or two, so we
+	// can just keep a single timeout context around for all calls.
+	ctxb := context.Background()
+	ctxt, cancel := context.WithTimeout(ctxb, defaultTimeout)
+	defer cancel()
+
+	// Derive a key we can use for signing.
+	keyDesc, internalKey, fullDerivationPath := deriveInternalKey(
+		ctxt, t, alice,
+	)
+
+	// Let's create a taproot script output now. This is a hash lock with a
+	// simple preimage of "foobar".
+	leaf1 := testScriptSchnorrSig(t.t, internalKey)
+
+	rootHash := leaf1.TapHash()
+	taprootKey := txscript.ComputeTaprootOutputKey(internalKey, rootHash[:])
+	tapScriptAddr, err := btcutil.NewAddressTaproot(
+		schnorr.SerializePubKey(taprootKey), harnessNetParams,
+	)
+	require.NoError(t.t, err)
+	p2trPkScript, err := txscript.PayToAddrScript(tapScriptAddr)
+	require.NoError(t.t, err)
+
+	// We need to assemble the control block to be able to spend through the
+	// script path.
+	tapscript := input.TapscriptPartialReveal(internalKey, leaf1, nil)
+	controlBlockBytes, err := tapscript.ControlBlock.ToBytes()
```

Comment:
Ok but in this case we actually construct the _valid_ control block?

Reply:
Yes. I know this is probably a bit confusing, so let me explain:

We have to differentiate between a call that comes in through the `SignOutputRaw` RPC in the remote signing case (where it's `SignOutputRaw(watch_only) -> SignPsbt(remote_signer)`) and a call to the `SignPsbt` call directly by the user/client (where it's `SignPsbt(watch_only) -> SignPsbt(remote_signer)`).

In the first case, we don't have all the information we need for a control block, because the `SignOutputRaw` RPC doesn't have a field for the internal key. _BUT_ the `SignOutputRaw` isn't expected to return a full and valid witness stack, only _the signature itself_. So we don't _need_ a fully valid control block, only a formally valid one. The signer doesn't care anyway since it doesn't look at the control block.

We only care about the correct/valid control block when we assemble our witness stack manually or go through the `FinalizePsbt` call. And this answers your other question about the roles: A finalizer is supposed to be able to assemble the complete witness stack from the signatures, leaf scripts and control blocks.

Does this all make sense? If yes, do you have an idea where/how I can/should document this for future questions?

Reply:
> So we don't need a fully valid control block, only a formally valid one. The signer doesn't care anyway since it doesn't look at the control block.

Ahh, ok this makes sense. Thanks for the explanation!

Reply:
> Does this all make sense? If yes, do you have an idea where/how I can/should document this for future questions?

So I think the main challenge here is that the remote signer proxy also uses the same set of protos as the normal user-facing `signrpc` system...

Thinking about it a bit more, afaict this doesn't actually affect the end consumers of the vanilla `signrpc` API and is mostly an internal quirk that wouldn't really pop up anywhere else other than code review. I think for the PSBT API side of things, we just need to make sure we're returning clear errors if one attempts to sign w/o providing all the required PSBT fields.